node_modules/gh-orgunit/lib/internal/dao.js

```diff
@@ -415,3 +416,377 @@ var getOrgUnitCalendar = module.exports.getOrgUnitCalendar = function(orgUnit, s
+ * @param  {Function}   callback                    Standard callback function
+ * @param  {Object}     callback.err                An error object, if any
+ * @param  {Number[]}   callback.allOrgUnitIds      The ids of the organisational units that are direct children of the given set of organisational unit ids
+ * @api private
+ */
+var _getDirectChildOrgUnits = function(parentOrgUnitIds, callback) {
+    var options = {
+        'attributes': ['id'],
+        'where': {
+            'ParentId': parentOrgUnitIds
+        }
+    };
+    DB.OrgUnit.findAll(options).complete(function(err, childOrgUnits) {
+        if (err) {
+            return callback(err);
+        }
+
+        var ids = _.map(childOrgUnits, function(childOrgUnit) {
+            return childOrgUnit.id;
+        });
+        return callback(null, ids);
+    });
+};
+
+/**
+ * Export the organisational units, event series and events under an organisational unit to CSV
+ *
+ * @param  {Number}         rootId              The id of the root organisational unit in the tree of organisational units, event series and events
+ * @param  {OrgUnit[]}      orgUnits            All the organisational units that are in the tree. Each organisational unit should have the event series and events available on its object that are under it
+ * @param  {Function}       callback            Standard callback function
+ * @param  {Object}         callback.err        An error object, if any
+ * @param  {String}         callback.data       The exported organisational units, event series and events as CSV
+ * @see gh-orgunit/lib/rest.js for more information
+ * @api private
+ */
+var _exportOrgUnitCSV = function(rootId, orgUnits, callback) {
+    // Get the simplified tree by exporting it to JSON first
+    _exportOrgUnitJSON(rootId, orgUnits, function(err, exportedOrgUnit) {
+        if (err) {
+            return callback(err);
+        }
+
+        var columns = _getExportCSVColumns(exportedOrgUnit);
+        if (!columns) {
+            return callback({'code': 400, 'msg': 'Tried to export an organisational unit that has children with the same type of different levels'});
+        }
+
+        var records = _getExportCSVRecords(exportedOrgUnit);
+
+        // Stringify the records into CSV
+        var options = {
+            'header': true,
+            'columns': columns
+        };
+        return csv.stringify(records, options, callback);
+    });
+};
+
+/**
+ * Get the column headers for a CSV export of an organisational unit tree. If the tree contained
+ * the same type of organisational unit on different levels, this function will return `null`.
+ * This indicates that the specific tree is malformed and cannot be exported to CSV
+ *
+ * @param  {Object}         exportedOrgUnit     A tree as generated by `exportOrgUnitJSON`
+ * @return {String[]}                           The column headers for a CSV export
+ * @api private
+ */
+var _getExportCSVColumns = function(exportedOrgUnit) {
```

Comment:
One of the reasons to not support the same type on different levels was so we could block it on import as well. With the `import` endpoint, we're already giving quite a bit of control away as to how people can mess up the data structure.

We could ditch using types for the column headers and just use the level number as a header I suppose.
A correct tree could become:
```
 + course (1)
     + part (2)
         + module (3)
         + module (4)
     + part (5)
         + module (6)
         + module (7)
 ```

could become:
```
0_id, 0_type, 0_parent, 0_.., 1_id, 1_type, 1_parent, 1_, ..., 2_id, 2_type, 2_parent, ...
1, course, , ..,            , 2, part, 1,                    , 3, module, 2
1, course, , ..,            , 2, part, 1,                    , 4, module, 2
1, course, , ..,            , 5, part, 1,                    , 6, module, 5
1, course, , ..,            , 5, part, 1,                    , 7, module, 5
```

A malformed tree:
```
 + course (1)
     + course (2)
         + module (3)
         + module (4)
     + part (5)
         + module (6)
         + module (7)
```
could become:
```
0_id, 0_type, 0_parent, 0_.., 1_id, 1_type, 1_parent, 1_, ..., 2_id, 2_type, 2_parent, ...
1, course, , ..,            , 2, course, 1,                  , 3, module, 2
1, course, , ..,            , 2, course, 1,                  , 4, module, 2
1, course, , ..,            , 5, part, 1,                    , 6, module, 5
1, course, , ..,            , 5, part, 1,                    , 7, module, 5
```

That means the column headers get pretty meaningless for people who are looking at them in excel and will use the export, manipulate them a bit and re-import it. Maybe we can try to use proper "type labels" when possible and fall back on the "level-index labels" when there's a malformed tree.


On the other hand, maybe we could try to be not so generic and make the import/export logic a timetable only feature and impose the necessary course / subject / part / module structure on it?